File: src/validation/validator.py
"""
Validação de qualidade após a limpeza.

Regra de negócio: antes de considerar um CSV "pronto para o Power BI",
o sistema roda checagens automáticas e reporta claramente se algo está
fora do esperado — em vez de simplesmente exportar e deixar o problema
ser descoberto só quando o painel já estiver publicado.
"""

import pandas as pd


class DataValidator:

    def __init__(self, logger=None):
        self.logger = logger
        self.alertas = []

    def _log(self, mensagem: str, nivel: str = "info"):
        if self.logger:
            getattr(self.logger, nivel)(mensagem)
        else:
            print(f"[VALIDAÇÃO] {mensagem}")

    def validar(self, df: pd.DataFrame, colunas_obrigatorias: list = None,
                linhas_minimas_esperadas: int = None) -> dict:
        """
        Roda o conjunto de checagens e devolve um dicionário com o resultado.
        Não interrompe a execução — reporta os problemas para decisão humana.
        """
        self.alertas = []
        resultado = {"aprovado": True, "alertas": []}

        # 1. Colunas obrigatórias presentes
        if colunas_obrigatorias:
            faltantes = [c for c in colunas_obrigatorias if c not in df.columns]
            if faltantes:
                self._registrar_alerta(f"Colunas obrigatórias ausentes: {faltantes}", resultado)

        # 2. Volume mínimo de linhas
        if linhas_minimas_esperadas and len(df) < linhas_minimas_esperadas:
            self._registrar_alerta(
                f"Volume de linhas abaixo do esperado: {len(df)} (esperado no mínimo {linhas_minimas_esperadas}).",
                resultado,
            )

        # 3. Colunas totalmente vazias (sinal de erro de leitura/mapeamento)
        colunas_vazias = [c for c in df.columns if df[c].isna().all()]
        if colunas_vazias:
            self._registrar_alerta(f"Colunas totalmente vazias (revisar mapeamento): {colunas_vazias}", resultado)

        # 4. Colunas "Unnamed" (sinal clássico de cabeçalho mal lido)
        colunas_unnamed = [c for c in df.columns if str(c).startswith("Unnamed")]
        if colunas_unnamed:
            self._registrar_alerta(
                f"Colunas 'Unnamed' detectadas (provável linha de título incorreta no arquivo original): {colunas_unnamed}",
                resultado,
            )

        # 5. Linhas totalmente duplicadas remanescentes
        duplicatas = df.duplicated().sum()
        if duplicatas > 0:
            self._registrar_alerta(f"{duplicatas} linhas totalmente duplicadas ainda presentes na base.", resultado)

        if resultado["aprovado"]:
            self._log("Validação concluída: nenhum problema encontrado.")
        else:
            self._log(f"Validação concluída com {len(resultado['alertas'])} alerta(s). Revisar antes de publicar.", nivel="warning")

        return resultado

    def _registrar_alerta(self, mensagem: str, resultado: dict):
        resultado["aprovado"] = False
        resultado["alertas"].append(mensagem)
        self._log(f"⚠ {mensagem}", nivel="warning")

File: src/validation/validator.ts
import { DataFrame } from "../types";
import { Logger } from "../utils/logger";

export class DataValidator {
  private logger?: Logger;
  public alertas: string[] = [];

  constructor(logger?: Logger) {
    this.logger = logger;
  }

  private _log(msg: string, nivel: "info" | "warning" | "error" = "info") {
    if (this.logger) {
      this.logger[nivel](msg);
    } else {
      console.log(`[VALIDAÇÃO] ${msg}`);
    }
  }

  public validar(
    df: DataFrame,
    colunasObrigatorias?: string[],
    linhasMinimasEsperadas?: number
  ): { aprovado: boolean; alertas: string[] } {
    this.alertas = [];
    const resultado = { aprovado: true, alertas: [] as string[] };

    // 1. Colunas obrigatórias
    if (colunasObrigatorias && colunasObrigatorias.length > 0) {
      const faltantes = colunasObrigatorias.filter((c) => !df.columns.includes(c));
      if (faltantes.length > 0) {
        this._registrarAlerta(`Colunas obrigatórias ausentes: ${faltantes.join(", ")}`, resultado);
      }
    }

    // 2. Volume mínimo de linhas
    if (linhasMinimasEsperadas && df.data.length < linhasMinimasEsperadas) {
      this._registrarAlerta(
        `Volume de linhas abaixo do esperado: ${df.data.length} (esperado no mínimo ${linhasMinimasEsperadas}).`,
        resultado
      );
    }

    // 3. Colunas totalmente vazias
    const total = df.data.length;
    if (total > 0) {
      const colunasVazias = df.columns.filter((col) => {
        return df.data.every((r) => r[col] === null || r[col] === undefined || r[col] === "");
      });
      if (colunasVazias.length > 0) {
        this._registrarAlerta(`Colunas totalmente vazias (revisar mapeamento): ${colunasVazias.join(", ")}`, resultado);
      }
    }

    // 4. Colunas "Unnamed"
    const colunasUnnamed = df.columns.filter((c) => c.startsWith("Unnamed") || c.startsWith("__EMPTY"));
    if (colunasUnnamed.length > 0) {
      this._registrarAlerta(
        `Colunas 'Unnamed' detectadas (provável linha de título incorreta no arquivo original): ${colunasUnnamed.join(", ")}`,
        resultado
      );
    }

    // 5. Linhas totalmente duplicadas
    if (df.data.length > 0) {
      const seen = new Set<string>();
      let dups = 0;
      for (const r of df.data) {
        const k = df.columns.map((c) => String(r[c] ?? "")).join("|||");
        if (seen.has(k)) dups++;
        else seen.add(k);
      }
      if (dups > 0) {
        this._registrarAlerta(`${dups} linhas totalmente duplicadas ainda presentes na base.`, resultado);
      }
    }

    if (resultado.aprovado) {
      this._log("Validação concluída: nenhum problema encontrado.");
    } else {
      this._log(`Validação concluída com ${resultado.alertas.length} alerta(s). Revisar antes de publicar.`, "warning");
    }

    return resultado;
  }

  private _registrarAlerta(mensagem: string, resultado: { aprovado: boolean; alertas: string[] }) {
    resultado.aprovado = false;
    resultado.alertas.push(mensagem);
    this.alertas.push(mensagem);
    this._log(`⚠ ${mensagem}`, "warning");
  }
}
